Components/Optimizers/PreconditionedGradientDescent/itkAdaptiveStochasticPreconditionedGradientDescentOptimizer.h

```diff
@@ -0,0 +1,160 @@
+/*=========================================================================
+ *
+ *  Copyright UMC Utrecht and contributors
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *=========================================================================*/
+#ifndef __itkAdaptiveStochasticPreconditionedGradientDescentOptimizer_h
+#define __itkAdaptiveStochasticPreconditionedGradientDescentOptimizer_h
+
+#include "itkStochasticPreconditionedGradientDescentOptimizer.h"
+
+#include "vnl/vnl_math.h"
+#include "itkSigmoidImageFilter.h"
+
+namespace itk
+{
+/**
+ * \class AdaptiveStochasticPreconditionedGradientDescentOptimizer
+ * \brief This class implements a gradient descent optimizer with adaptive gain.
+ *
+ * If \f$C(x)\f$ is a cost function that has to be minimized, the following iterative
+ * algorithm is used to find the optimal parameters \f$x\f$:
+ *
+ *     \f[ x(k+1) = x(k) - a(t_k) P dC/dx \f]
+ *
+ * The gain \f$a(t_k)\f$ at each iteration \f$k\f$ is defined by:
+ *
+ *     \f[ a(t_k) =  a / (A + t_k + 1)^alpha \f].
+ *
+ * And the time \f$t_k\f$ is updated according to:
+ *
+ *     \f[ t_{k+1} = [ t_k + sigmoid( -g_k^T P g_{k-1} ) ]^+ \f]
+ *
+ * where \f$g_k\f$ equals \f$dC/dx\f$ at iteration \f$k\f$.
+ * For \f$t_0\f$ the InitialTime is used, which is defined in the
+ * the superclass (StandardGradientDescentOptimizer). Whereas in the
+ * superclass this parameter is superfluous, in this class it makes sense.
+ *
+ * This method is described in the following references:
+ *
```

Comment:
Qiao Y, Lelieveldt BP, Staring M. An efficient preconditioner for stochastic gradient descent optimization of image registration. IEEE transactions on medical imaging. 2019 Feb 11.